src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = { ...state };
+  const transformState = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        transformState.push({ ...newState });
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          if (newState[key] !== undefined) {
+            delete newState[key];
+          }
+        }
+        transformState.push({ ...newState });
+        break;
+
+      case 'clear':
+        for (const key in newState) {
+          delete newState[key];
```

Comment:
you can assign an empty object for newState -> avoid excessive loop for better readability